src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -0,0 +1,208 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Helper type for building dynamically-sized arrays while minimizing allocations and copying.
+    /// </summary>
+    internal struct LargeArrayBuilder<T>
+    {
+        private const int StartingCapacity = 4;
+        private const int ResizeLimit = 32;
+
+        private T[] _first;                 // The first buffer we store items in. Resized until ResizeLimit.
+        private ArrayBuilder<T[]> _buffers; // After ResizeLimit * 2, we store previous buffers we've filled out here.
+        private T[] _current;               // Current buffer we're reading into. Is _first <= ResizeLimit.
+        private int _index;                 // Index into the current buffer.
+        private int _count;                 // Count of all of the items in this builder.
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public LargeArrayBuilder(bool initialize) : this()
+        {
+            // This is a workaround for C# not having parameterless struct constructors yet.
+            // Once it gets them, replace this with a parameterless constructor.
+            Debug.Assert(initialize);
+
+            _first = _current = Array.Empty<T>();
+        }
+
+        /// <summary>
+        /// Gets the number of items added to the builder.
+        /// </summary>
+        public int Count => _count;
+
+        /// <summary>
+        /// Adds an item to this builder.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <remarks>
+        /// Use <see cref="Add"/> if adding to the builder is a bottleneck for your use case.
+        /// Otherwise, use <see cref="SlowAdd"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add(T item)
```

Comment:
@stephentoub, good. My percentages may have been skewed a bit since I was comparing the smaller times in relation to the larger times instead of vice versa (1100 is a 20% regression from 900, 900 is an 18% improvement from 1100) so I will have to be careful about that in the future.